### PR TITLE
[KEYCLOAK-8104] Keycloak SAML Adapter does not support clockSkew conf…

### PR DESCRIPTION
**Pull request by steevebtib**
…iguration

### CODE REVIEW

File: adapters/saml/core/src/main/java/org/keycloak/adapters/saml/config/parsers/DeploymentBuilder.java [hunk 1]
```diff
@@ -214,6 +220,25 @@ public SamlDeployment build(InputStream xml, ResourceLoader resourceLoader) thro
         return deployment;
     }
 
+    private int convertClockSkewInMillis(int duration, TimeUnit unit) {
```
Comment (vramik) [inline adapters/saml/core/src/main/java/org/keycloak/adapters/saml/config/parsers/DeploymentBuilder.java]: I'd suggest to just cut the rest from nanoseconds and microsecods, It'd simplify the method. We also doesn't need `long` here, `int` would be sufficient. And the INFO log message with value in ms should be added.
```suggestion
    private int convertClockSkewInMillis(int duration, TimeUnit unit) {
        int durationMillis = (int) unit.toMillis(duration);
        switch (unit) {
            case NANOSECONDS:
            case MICROSECONDS:
                log.warn("Clock skew value will be rounded down.");
            default:
                log.info("Clock skew set to " + durationMillis + "ms.");
        }
        return durationMillis;
    }
```

**Review by vramik — CHANGES_REQUESTED** [review]
@steevebtib thank you a lot for the PR. I've noticed there is no integration tests. Would you have time to write some? I can help you with those if needed or I can add some myself. 

Otherwise it looks good to me, I've got few minor things.

File: adapters/saml/core/src/main/java/org/keycloak/adapters/saml/config/parsers/DeploymentBuilder.java [hunk 2]
```diff
@@ -170,7 +168,15 @@ public SamlDeployment build(InputStream xml, ResourceLoader resourceLoader) thro
         sso.setRequestBindingUrl(idp.getSingleSignOnService().getBindingUrl());
         if (idp.getSingleSignOnService().getResponseBinding() != null) {
             sso.setResponseBinding(SamlDeployment.Binding.parseBinding(
-                idp.getSingleSignOnService().getResponseBinding()));
+                    idp.getSingleSignOnService().getResponseBinding()));
+        }
+        if (idp.getAllowedClockSkew() != null) {
+            defaultIDP.setAllowedClockSkew(convertClockSkewInMillis(idp.getAllowedClockSkew(), idp.getAllowedClockSkewUnit()));
+        }
+        sso.setRequestBinding(SamlDeployment.Binding.parseBinding(sp.getIdp().getSingleSignOnService().getRequestBinding()));
```
[the remark this answers is not in the document]
Comment (steevebtib) [reply adapters/saml/core/src/main/java/org/keycloak/adapters/saml/config/parsers/DeploymentBuilder.java]: in fact, this was related to a previous commit that I fetched after my initial pull, I will see how I can remove my updates on those lines

**Comment by steevebtib** [comment]
I can try to write integration test, could you please explain how to proceed?

**Comment by vramik** [comment]
@steevebtib 

> I can try to write integration test, could you please explain how to proceed? 

I'll create sample test and send it to you. Then you can add more tests according this one, does that work for you? Maybe we could create another task for tests, I'm not sure, @hmlnarik wdyt?

> I guess that I should amend and force push for the requested modifications right?

Exactly, thank you

**Comment by steevebtib** [comment]
@vramik 
> @steevebtib
> 
> > I can try to write integration test, could you please explain how to proceed?
> 
> I'll create sample test and send it to you. Then you can add more tests according this one, does that work for you? Maybe we could create another task for tests, I'm not sure, @hmlnarik wdyt?
That's ok for me, thank you for your feedback, I will wait for you then.

**Comment by hmlnarik** [comment]
> I'll create sample test and send it to you. Then you can add more tests according this one, does that work for you? Maybe we could create another task for tests, I'm not sure, @hmlnarik wdyt?

The tests need to be part of the same task, please keep it in the same PR.

Side note - if you prefer, the PR can contain several commit while in progress. These would need to be squashed into a single one before final review and merge.

**Comment by steeveborc** [comment]
@vramik 
Do you confirm that the integration tests for this use case go to package **org.keycloak.testsuite.saml** inside of project **integration-arquillian-tests-base** ? This is the place I have found by browsing your source and documentation

**Comment by vramik** [comment]
@steeveborc I've opened [PR](https://github.com/steevebtib/keycloak/pull/1) to your fork with example test, pls read comments there.

**Comment by hmlnarik** [comment]
@steevebtib Would you please have time to continue work on this PR?

**Comment by steevebtib** [comment]
@hmlnarik I have implemented the changes, I will try to implement integration test tomorrow as Vlasta has provided an example.

**Comment by hmlnarik** [comment]
@steevebtib Before submitting the next round of changes, please rebase to latest upstream master

**Comment by steevebtib** [comment]
@hmlnarik ok I will do

**Comment by vramik** [comment]
@steevebtib to make the travis pass this [commit](https://github.com/keycloak/keycloak/pull/6050#issuecomment-495109831) needs to also backported, thanks

**Comment by steevebtib** [comment]
@vramik yes I have fetched your latest update, and then I will rebase on latest upstream master

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/SAMLClockSkewAdapterTest.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+                    }).build().executeAndTransform(resp -> EntityUtils.toString(resp.getEntity()));
+
+            Assert.assertThat(resultPage, matcher);
+        } finally {
+            setAdapterAndServerTimeOffset(0);
+        }
+    }
+
+    @Test
+    public void testErrorExample() throws Exception {
```
Comment (vramik) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/SAMLClockSkewAdapterTest.java]: nit: the test method name was meant to be changed, I was sending it as an example for you to make it easier for you add another tests.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/SAMLClockSkewAdapterTest.java [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+                ));
+            }
+
+        } finally {
+            deployer.undeploy(DEPLOYMENT_NAME_3_SEC);
+        }
+    }
+
+    @Test
+    public void testSuccessExample() throws Exception {
```
Comment (vramik) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/SAMLClockSkewAdapterTest.java]: dtto

**Comment by vramik** [comment]
@steevebtib thanks, I've added two more comments regarding test names, but those are rather nits. Thanks

**Comment by vramik** [comment]
Hello @steevebtib, would you have possibility to continue with this one?

**Comment by steevebtib** [comment]
Hello @vramik, I should have more time tomorrow to continue on this, I have merged your latest updates, but not updated the tests yet either rebased on upstream

**Comment by steevebtib** [comment]
Hello @vramik, I have modified the tests and added a new one, and got all the latest changes, I am just checking again all tests

**Comment by hmlnarik** [comment]
@steevebtib Feel free to submit the changes here, the automation will do the testing for all the app servers for you. Should there be any issue, we will be happy to help, but I believe that it will pass just fine.

Also please consider including the following line in the commit message since @vramik developed the testsuite support for clock skew testing delivered in this PR:
```
Co-Authored-By: vramik <vramik@redhat.com>
```

**Comment by steevebtib** [comment]
@hmlnarik I have done all the requested changes and force pushed, I hope it will be fine for you.

**Comment by hmlnarik** [comment]
@steevebtib Thank you, overall the PR looks good. It is now being checked.

**Comment by vramik** [comment]
@steevebtib I've opened https://github.com/steevebtib/keycloak/pull/3 with fix for tomcat and jetty app servers, once that one is merged and rebased to latest upstream master, it's good to go in. Thanks

**Comment by hmlnarik** [comment]
@steevebtib Could you please merge the steevebtib/keycloak#3 and squash the commits? Once this is ready, the PR can be merged.

**Comment by steevebtib** [comment]
@hmlnarik Hi, sorry I was too busy the last days, I didn’t have time to check but I will do tomorrow

**Review by vramik — APPROVED** [review]

**Review by hmlnarik — APPROVED** [review]
Thanks @steevebtib

**Pull request merged**